trace-forward/src/Trace/Forward/Queue.hs

```diff
@@ -0,0 +1,58 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Trace.Forward.Queue
+  ( readItems
+  , getTraceObjects
+  ) where
+
+import           Control.Concurrent.STM (STM, atomically, retry)
+import           Control.Concurrent.STM.TBQueue (TBQueue, tryReadTBQueue)
+import qualified Data.List.NonEmpty as NE
+import           Data.Word (Word16)
+
+import           Trace.Forward.Configuration (ForwarderConfiguration (..))
+import qualified Trace.Forward.Protocol.Forwarder as Forwarder
+import           Trace.Forward.Protocol.Type
+
+readItems
+  :: ForwarderConfiguration lo -- ^ The forwarder configuration.
+  -> TBQueue lo                -- ^ The queue we will read 'TraceObject's from.
+  -> Forwarder.TraceForwarder lo IO ()
+readItems config@ForwarderConfiguration{..} loQueue =
+  Forwarder.TraceForwarder
+    { Forwarder.recvMsgNodeInfoRequest = do
+        reply <- getNodeInfo
+        return (reply, readItems config loQueue)
+    , Forwarder.recvMsgRequest = \blocking (GetTraceObjects n) -> do
+        replyList <-
+          case blocking of
+            TokBlocking -> do
+              objs <- atomically $ getNTraceObjects n loQueue >>= \case
+                []     -> retry -- No 'TraceObject's yet, just wait...
+                (x:xs) -> return $ x NE.:| xs
+              return $ BlockingReply objs
+            TokNonBlocking -> do
+              objs <- atomically $ getNTraceObjects n loQueue
+              return $ NonBlockingReply objs
+        return (replyList, readItems config loQueue)
+    , Forwarder.recvMsgDone = return ()
+    }
+
+-- | Returns at most N 'TraceObject's from the queue.
+getNTraceObjects
+  :: Word16
+  -> TBQueue lo
+  -> STM [lo]
+getNTraceObjects 0 _ = return []
+getNTraceObjects n loQueue =
+  tryReadTBQueue loQueue >>= \case
+    Just lo' -> (:) lo' <$> getNTraceObjects (n - 1) loQueue
+    Nothing  -> getNTraceObjects 0 loQueue
```

Comment:
I think this will be clearer:
```suggestion
    Nothing  -> return []
```

Reply:
Anther implementation is `catMaybes <$> replicateM n (tryReadTBQueue q)`. It's also non blocking but it does not shortcut if it's the queue is shallow. I would refactor your implementation into:
```
getNTraceObjects n0 q = go n0
  where
    go 0 = return  []
    go n = tryReadTBQueue q >>= \case
       Just lo   -> (lo :) <$> go (pred n)
       Nothing -> return []
```